### PR TITLE
Assign Tag on Product Purchase

### PR DESCRIPTION
**Pull request by n7studios**
## Summary

Adds requested functionality to tag a subscriber based on the MemberMouse product purchased.

Screenshot 2024-07-02 at 15 54 50

## Testing

- `ProductTagCest`: Tests that members are subscribed and tagged in ConvertKit when configured to do so in the Plugin's settings

## Checklist

* [x] I have [written a test](TESTING.md#writing-an-acceptance-test) and included it in this PR
* [x] I have [run all tests](TESTING.md#run-tests) and they pass
* [x] The code passes when [running the PHP CodeSniffer](TESTING.md#run-php-codesniffer)
* [x] Code meets [WordPress Coding Standards](DEVELOPMENT.md#coding-standards) for PHP, HTML, CSS and JS
* [x] [Security and Sanitization](DEVELOPMENT.md#security-and-sanitization) requirements have been followed
* [x] I have assigned a reviewer or two to review this PR (if you're not sure who to assign, we can do this step for you)

### CODE REVIEW

File: admin/class-convertkit-mm-admin.php [hunk 1]
```diff
@@ -142,7 +143,33 @@ public function register_settings() {
 						'tags' => $tags,
 					)
 				);
+			}
+		}
 
+		// Output product to tag mappings.
+		if ( empty( $products ) ) {
+			add_settings_field(
+				'convertkit-empty-mapping',
```
Comment (noelherrick) [inline admin/class-convertkit-mm-admin.php:152]: Should this be a constant?

File: admin/class-convertkit-mm-admin.php [hunk 2]
```diff
@@ -142,7 +143,33 @@ public function register_settings() {
...
+				apply_filters( $this->plugin_name . '_display_convertkit_mapping_product', esc_html__( 'Mapping', 'convertkit-mm' ) ),
+				array( $this, 'display_options_empty_mapping_products' ),
+				$this->plugin_name,
+				$this->plugin_name . '-ck-mapping'
+			);
+
+		} else {
+			foreach ( $products as $key => $name ) {
+				add_settings_field(
+					'convertkit-mapping-product-' . $key,
```
Comment (noelherrick) [inline admin/class-convertkit-mm-admin.php:162]: This is another one that's probably a good candidate for a constant (and not an env var)

File: tests/acceptance/general/ProductTagCest.php [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+		$I->fillField('mm_field_last_name', 'Last');
+		$I->fillField('mm_field_email', $emailAddress);
+		$I->fillField('mm_field_password', '12345678');
+		$I->fillField('mm_field_phone', '12345678');
+		$I->fillField('mm_field_cc_number', '4242424242424242');
+		$I->fillField('mm_field_cc_cvv', '123');
+		$I->selectOption('mm_field_cc_exp_year', '2038');
+		$I->fillField('mm_field_billing_address', '123 Main Street');
+		$I->fillField('mm_field_billing_city', 'Nashville');
+		$I->selectOption('#mm_field_billing_state_dd', 'Tennessee');
```
Comment (noelherrick) [inline tests/acceptance/general/ProductTagCest.php:77]: Just verifying the octothorp/hash mark is intended.

**Review by noelherrick — APPROVED** [review]

File: tests/acceptance/general/ProductTagCest.php (hunk 3 above)
Comment (n7studios) [reply tests/acceptance/general/ProductTagCest.php:77]: Yes, this is intended. The field doesn't have a `name` attribute in MemberMouse, for some reason.

File: admin/class-convertkit-mm-admin.php (hunk 2 above)
Comment (n7studios) [reply admin/class-convertkit-mm-admin.php:162]: Agreed - I'd like to move settings to their own class with a constant as we do in the main ConvertKit plugin, and will do this in another PR.

File: admin/class-convertkit-mm-admin.php (hunk 1 above)
Comment (n7studios) [reply admin/class-convertkit-mm-admin.php:152]: Agreed - I'd like to move settings to their own class with a constant as we do in the main ConvertKit plugin, and will do this in another PR.

**Pull request merged**